test/metabase/query_processor/middleware/permissions_test.clj

```diff
@@ -411,3 +411,46 @@
                  :type                  :native
                  :qp/compiled-from-mbql {:source-table (u/the-id table)}
                  :native                {:query "SELECT * FROM VENUES"}}))))))))
+
+(deftest e2e-native-query-source-card-id-join-perms-test
+  (testing "Make sure that a native source card joined to an MBQL query checks card read perms rather than full native access"
+    (mt/with-temp [:model/User {user-id :id} {}
+                   :model/Card {card-id :id} {:dataset_query {:database (mt/id)
+                                                              :type :native
+                                                              :native {:query "SELECT * FROM venues"}}}]
+      (mt/with-no-data-perms-for-all-users!
+        (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/create-queries :query-builder)
```

Comment:
IIUC these two tests verify that

1. (This test) You *can* run the query if you have `:query-builder` perms and read perms for the Card even if the Card query is native (fixes the bug)
2. (Test below) You *cannot* run the query without `:query-builder-and-native` perms if query is actually a native query, even if you try to override it by passing in `:qp/stage-is-from-source-card` yourself.

Does it make sense to add an additional case similar to (1) where the user *doesn't* have perms for the Card, or is that covered elsewhere?

Reply:
added a new test — `e2e-nested-source-card-full-permissions-test` which is a copy of `e2e-nested-source-card-test`, but where the nested card isn't in an accessible collection. I think this now covers all the possible cases.